ipn/localapi/localapi.go

```diff
@@ -1291,14 +1291,16 @@ func (h *Handler) serveWatchIPNBus(w http.ResponseWriter, r *http.Request) {
 	}
 
 	w.Header().Set("Content-Type", "application/json")
+	buf := bytes.NewBuffer(nil)
 	ctx := r.Context()
 	h.b.WatchNotifications(ctx, mask, f.Flush, func(roNotify *ipn.Notify) (keepGoing bool) {
-		js, err := json.Marshal(roNotify)
+		buf.Reset()
```

Comment:
Oh, buf.Reset first, not after the w.Write. Yeah, so this should actually retain more memory as you wrote.

But that seems like the opposite direction to move in. The notifications here can be megabytes of JSON. We don't want to retain megabytes when idle.

Reply:
We have a tiered buffer pool in corp, should we move that to OSS as well?

Reply:
In corp we also do:

```go
// jenc is a json.Encode + bytes.Buffer pair wired up to be reused in a pool.
type jenc struct {
	buf bytes.Buffer
	enc *json.Encoder
}

var jencPool = &sync.Pool{New: func() interface{} {
	je := new(jenc)
	je.enc = json.NewEncoder(&je.buf)
	return je
}}
```

Reply:
As I noted, this trades retained memory in exchange for reduced allocations. This buffer corresponds to a single client connection to localapi's watch endpoint. In practice (at least on iOS) we have only one of these, so pooling doesn't buy us anything, and versus pooling this has the advantage of not creating multiple potentially large buffers.

That said, slacking with @bradfitz he's starting to convince me that we should be able to set up the GC so that garbage should never push us over our memory limit, so I'll focus on that.